py/plugins/vertex-ai/src/genkit/plugins/vertex_ai/gemini.py

```diff
@@ -92,41 +102,89 @@ def __init__(self, version: str):
+                    parts.append(genai.Part.from_text(part.root.text))
+                elif isinstance(part.root, MediaPart):
+                    if not self.is_multimode():
+                        LOG.error(
+                            f'The model {self._version} does not'
+                            f' support multimode input'
+                        )
+                        continue
+                    parts.append(
+                        genai.Part.from_uri(
+                            mime_type=part.root.media.content_type,
+                            uri=part.root.media.url
+                        )
+                    )
+                elif isinstance(part.root, DataPart | CustomPart):
+                    LOG.warning(f'The type {type(part.root)} is'
+                                f' not yet implemented.')
+                else:
+                    LOG.error('The type is not supported')
+            messages.append(genai.Content(role=message.role.value, parts=parts))
+        return messages
+
     @property
-    def gemini_model(self) -> GenerativeModel:
+    def gemini_model(self) -> genai.GenerativeModel:
         """Get the Vertex AI GenerativeModel instance.
 
         Returns:
             A configured GenerativeModel instance for the specified version.
         """
-        return GenerativeModel(self._version)
+        return genai.GenerativeModel(self._version)
 
-    def handle_request(self, request: GenerateRequest) -> GenerateResponse:
+    def generate(
+        self, request: GenerateRequest, ctx: ActionRunContext
+    ) -> GenerateResponse | None:
         """Handle a generation request using the Gemini model.
 
         Args:
             request: The generation request containing messages and parameters.
+            ctx: additional context
 
         Returns:
             The model's response to the generation request.
         """
-        messages: list[Content] = []
-        for m in request.messages:
-            parts: list[Part] = []
-            for p in m.content:
-                if p.root.text is not None:
-                    parts.append(Part.from_text(p.root.text))
-                else:
-                    raise Exception('unsupported part type')
-            messages.append(Content(role=m.role.value, parts=parts))
-        response = self.gemini_model.generate_content(contents=messages)
-        return GenerateResponse(
-            message=Message(
-                role=Role.MODEL,
-                content=[TextPart(text=response.text)],
-            )
+
+        messages = self.build_messages(request)
+        response = self.gemini_model.generate_content(
+            contents=messages, stream=ctx.is_streaming
         )
 
+        if ctx.is_streaming:
+            for chunk in response:
+                ctx.send_chunk(
+                    GenerateResponseChunk(
+                        role=Role.MODEL,
+                        content=[TextPart(text=chunk.text)],
```

Comment:
please add a TODO here to address later -- must handle function/tool calling parts during streaming as well